[MRG+1] correct condition in decision tree construction

nelson-liu opened the pull request:
#### Reference Issue

Continuation of #7340 
#### What does this implement/fix? Explain your changes.

Removes an unused leaf condition in `_tree.pyx`.
#### Any other comments?

ping @jnothman, who I previously discussed this with

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/7441)

<!-- Reviewable:end -->

nelson-liu committed:
remove unused condition in decision tree construction

jnothman commented:
While I think these conditions are logically inappropriate, just in case I'm wrong, have you confirmed that they *are* never executed at least by the test suite?

nelson-liu commented:
just verified; in the original code, never is `weighted_n_node_samples < min_weight_leaf` True after the `is_leaf` conditional changed here.

jnothman commented:
I mean with something like printing a message if the condition is ever true
here

On 16 September 2016 at 07:12, Nelson Liu <notifications@github.com> wrote:

> just verified; in the original code, never is weighted_n_node_samples <
> min_weight_leaf True after the is_leaf conditional changed here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7441#issuecomment-247457062>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yo7jjjxDvHfGJuRipCQDXJ3Dqh-ks5qqbTOgaJpZM4J-TZK>
> .
>

nelson-liu commented:
yes, that's what i did (and nothing was printed). sorry for being unclear.

jnothman commented:
Great, thanks. LGTM

On 16 September 2016 at 08:46, Nelson Liu <notifications@github.com> wrote:

> yes, that's what i did (and nothing was printed). sorry for being unclear.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7441#issuecomment-247475999>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6zNg3-eqW__PJocp5O2LgZZVKYyhks5qqcqogaJpZM4J-TZK>
> .
>

amueller commented:
would be great to have @glouppe of @jmschrei look at this.

glouppe commented:
Hmmm can you explain why we should remove it? What if you make a test with `sample_weight` such that `sample_weight.sum() < min_weight_leaf` from the beginning?

jmschrei commented:
Is it possible the test suite just isn't comprehensive enough to test cases where this might be important? Can you create by hand a dataset where you would expect that the weight would cause a difference in the splits and confirm?

nelson-liu commented:
> Is it possible the test suite just isn't comprehensive enough to test cases where this might be important? 

yes, that's very possible. Are said "cases where this might be important" just when you have a dataset with `sample_weight` such that sample_weight.sum() < min_weight_leaf? or is there something else that i am missing?

jmschrei commented:
Yeah, but sample_weight for the samples in the current node, not the entire dataset. I haven't read through the test cases recently, but do we have a test where we should make a split on weighted data (not unweighted data), but don't because of min_weight_leaf?

jnothman commented:
What does it mean logically to say that something should be a leaf if its weight is so small that it violates the `min_weight_leaf` condition?

jnothman commented:
@glouppe:

> Hmmm can you explain why we should remove it? What if you make a test with sample_weight such that sample_weight.sum() < min_weight_leaf from the beginning?

This shouldn't be possible through the public interface, which sets `min_weight_fraction_leaf`, not `min_weight_leaf`, and requires the former to be at most `.5 * sample_weight.sum()`.

(If this isn't a valid change to make, then surely changing `<` to `<=` is required in these conditions.)

raghavrv commented:
IIRC Nothing would break even if you remove the whole set of conditions as the splitter will check it again.

The reason why this check is given here is to check and avoid entering the splitter at all if we know it is a leaf for sure. And I think the condition should instead be `weighted_n_node_samples < 2 * min_weight_leaf`...

> just verified; in the original code, never is `weighted_n_node_samples < min_weight_leaf` `True` after the `is_leaf` conditional changed here.

Where does "here" refer to? How did you verify?

The flow, IIUC, is as follows

Assume `min_weight_leaf=0.1` and `min_samples_leaf=1`
Assume a particular node's sample weights to be `[0.05, 0.05, 0.05, 0.05, 0.05]`
The `weighted_n_node_samples` for that node is now `0.25`

* Let that node be the one that is currently popped from stack.
* Since one can split the data into 2 parts having weights at right `[0.05, 0.05, 0.05] and left `[0.05, 0.05]`, it is allowed to enter the splitter.
* Assume the splitter splits it that way. Splitter tells us it is not a leaf and gives us the 2 partitions.
* The right/left partitions are stacked.
* The left is popped. Now since the cumulative weights for the left partition is 0.1, there is no way we can split the data without violating the `min_weight_leaf=0.1`.
* Hence the condition that you just removed (if corrected as mentioned above) marks it as leaf before it enters the splitter and saves us some computational time.

jnothman commented:
> And I think the condition should instead be `weighted_n_node_samples < 2 * min_weight_leaf`

I had originally thought that condition inappropriate (@nelson-liu had suggested it), but now I think you might be right. If that condition is true, there is no way to split the points at that node such that both splits satisfy `min_weight_leaf`, is there?

Yes, I now think this is appropriate as a fast path. However, this will change the fitted trees, so we should note that in what's new.

Thanks.

raghavrv commented:
No it won't change the fitted trees, as before the splitter used to mark it as leaf. Now after the corrected condition, it will not enter the splitter at all. And yes with the corrected condition in place there is no way to split it such that `min_weight_leaf` is not violated...

jnothman commented:
Yes, that will change the trees due to random_state.

On 22 September 2016 at 10:07, Raghav RV <notifications@github.com> wrote:

> No it won't change the fitted trees, as before the splitter used to mark
> it as leaf. Now after the corrected condition, it will not enter the
> splitter at all. And yes with the corrected condition in place there is no
> way to split it such that min_weight_leaf is not violated...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7441#issuecomment-248778457>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60j4jCAoADkX6G-eomh3qWVusDwuks5qscavgaJpZM4J-TZK>
> .
>

raghavrv commented:
Wow. Good catch! I totally missed that. Indeed the tree will change!

nelson-liu commented:
following up with this; should we change the condition to `weighted_n_node_samples < 2 * min_weight_leaf` as suggested by @raghavrv ?

jnothman commented:
I think so. (Sorry for changing my mind!)

On 29 September 2016 at 00:14, Nelson Liu <notifications@github.com> wrote:

> following up with this; should we change the condition to weighted_n_node_samples
> < 2 * min_weight_leaf as suggested by @raghavrv
> <https://github.com/raghavrv> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7441#issuecomment-250178832>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6y804A24W9B4d6ZJfPk_r9SRKhNcks5qunZbgaJpZM4J-TZK>
> .
>

nelson-liu committed:
edit is_leaf condition for min_weight_leaf

nelson-liu commented:
no need to be sorry, I'm glad we were able to come to a better solution. pushed the new change.

nelson-liu committed:
edit ordering of statements

jnothman commented on sklearn/tree/_tree.pyx:
```diff
@@ -218,8 +218,8 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
 
                 is_leaf = ((depth >= max_depth) or
```
Do you mind removing all the extraneous parentheses too? :/

nelson-liu replied:
yeah, sorry about that.

jnothman commented:
please change PR title

jnothman commented:
actually, I will, seeing as I'm adding a +1. LGTM

jnothman commented:
What I'd like to see is a what's new entries that says it's more efficient but trees using the parameter will change.

nelson-liu committed:
remove extra parens and add whats new

nelson-liu commented:
>What I'd like to see is a what's new entries that says it's more efficient but trees using the parameter will change.

@jnothman thanks for changing the title for me. I added a what's new entry and removed the extra parens.

raghavrv commented:
Thanks for the change. This looks good to me. I think it would be better if @glouppe gave his final +1 and merge :)

glouppe commented:
LGTM too! Merging

jnothman commented on doc/whats_new.rst:
```diff
@@ -20,6 +20,13 @@ New features
+   - Edited criterion for leaf nodes in decision tree criterion by declaring a
+     node as a leaf if the weighted number of samples at the node is less than
+     2 * the minimum weight specified to be at a node. This makes growth more
+     efficient, but trees using parameters that modify the weight at each leaf
```
I think this is a confusing way of stating it. Will fix it up in master.

nelson-liu replied:
thanks @jnothman

jnothman replied:
20d0b00
